pkg/checker.go

```diff
@@ -0,0 +1,170 @@
+package pkg
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"golang.org/x/net/html"
+)
+
+const (
+	SuccessColor = "\033[1;32m%s\033[0m"
+	ErrorColor   = "\033[1;31m%s\033[0m"
+	DebugColor   = "\033[0;36m%s\033[0m"
+)
+
+/*type URLStatus struct {
+	url    string
+	err    string
+	status int
+	index  int
+}*/
+
+type URLStatus struct {
+
+	// err
+	Err string `json:"err,omitempty"`
+
+	// index
+	Index int64 `json:"index,omitempty"`
+
+	// status
+	Status int64 `json:"status,omitempty"`
+
+	// url
+	URL string `json:"url,omitempty"`
+}
+
+func checkWebsite(website string) (string, error) {
+	var err error
+	websiteTries := [3]string{website, "https://" + website, "http://" + website}
+	for _, web := range websiteTries {
+		_, err = http.Get(web)
+		if err == nil {
+			website = web
+		}
+	}
+	return website, err
+}
+
+// get links inside a given website
+func GetLinks(website string) ([]string, error) {
+	var links []string
+
+	// check the status of the website before getting its links
+	res, err := http.Get(website)
+	if err != nil {
+		return links, err
+	}
+
+	body := res.Body
+
+	htmlTokenizer := html.NewTokenizer(body)
+	for {
+		token := htmlTokenizer.Next()
+
+		switch token {
+		case html.ErrorToken:
+			// remove duplicates
+			testDuplicate := make(map[string]bool)
+			uniqueLinks := []string{}
+			for _, item := range links {
+				if _, value := testDuplicate[item]; !value {
+					testDuplicate[item] = true
+					uniqueLinks = append(uniqueLinks, item)
+				}
+			}
+
+			return uniqueLinks, nil
+		case html.StartTagToken, html.EndTagToken:
+			token := htmlTokenizer.Token()
+			if "a" == token.Data {
+				for _, attr := range token.Attr {
+					if attr.Key == "href" {
+						links = append(links, attr.Val)
+					}
+
+				}
+			}
+
+		}
+	}
+}
+
+// TestLink tests the response of get request
+// It returns the status code and the error of the request
+func TestLink(website, url string) (int, error) {
+	if strings.HasPrefix(url, "/") {
+		url = website + url
+	} else if !strings.HasPrefix(url, "http") {
+		url = website + "/" + url
+	}
+
+	res, err := http.Get(url)
```

Comment:
if the link on the same website we will need to check and visit all of its link, but if it's not on the website (external link) we can just do a `HEAD` request if that works we can report the link as valid (without having to read its content)

Reply:
status code should be in 20x 30x